File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/ObservationDeserializer.java
/*
 * The contents of this file are subject to the OpenMRS Public License
 * Version 1.0 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://license.openmrs.org
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
 */

package org.openmrs.mobile.utilities;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import org.openmrs.mobile.models.Concept;
import org.openmrs.mobile.models.Observation;

import java.lang.reflect.Type;

public class ObservationDeserializer implements JsonDeserializer<Observation> {

    private static final String UUID_KEY = "uuid";
    private static final String DISPLAY_KEY = "display";
    private static final String VALUE_KEY = "value";

    @Override
    public Observation deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {

        JsonObject jsonObject = json.getAsJsonObject();

        Observation observation = new Observation();
        observation.setUuid(jsonObject.get(UUID_KEY).getAsString());
        observation.setDisplay(jsonObject.get(DISPLAY_KEY).getAsString());

        JsonElement conceptJson = jsonObject.get("concept");
        if (conceptJson != null && "Visit Diagnoses".equals(conceptJson.getAsJsonObject().get(DISPLAY_KEY).getAsString())) {
            JsonArray diagnosisDetailJSONArray = jsonObject.get("groupMembers").getAsJsonArray();
            for (int i = 0; i < diagnosisDetailJSONArray.size(); i++) {

                JsonObject diagnosisDetails = diagnosisDetailJSONArray.get(i).getAsJsonObject();
                String diagnosisDetail = diagnosisDetails.get("concept").getAsJsonObject().get(DISPLAY_KEY).getAsString();

                if ("Diagnosis order".equals(diagnosisDetail)) {
                    observation.setDiagnosisOrder(
                            diagnosisDetails.getAsJsonObject().get(VALUE_KEY).getAsJsonObject().get(DISPLAY_KEY).getAsString());
                } else if ("Diagnosis certainty".equals(diagnosisDetail)) {
                    observation.setDiagnosisCertanity(
                            diagnosisDetails.getAsJsonObject().get(VALUE_KEY).getAsJsonObject().get(DISPLAY_KEY).getAsString());
                } else {
                    try {
                        observation.setDiagnosisList(diagnosisDetails.getAsJsonObject().get(VALUE_KEY).getAsJsonObject().get(DISPLAY_KEY).getAsString());
                    }
                    catch (IllegalStateException e) {
                        observation.setDiagnosisList(diagnosisDetails.getAsJsonObject().get(VALUE_KEY).getAsString());
                    }
                }
            }
        } else if (conceptJson != null && "Text of encounter note".equals(conceptJson.getAsJsonObject().get(DISPLAY_KEY).getAsString())) {
            observation.setDiagnosisNote(jsonObject.getAsJsonObject().get(VALUE_KEY).getAsString());
        }
        if (conceptJson != null) {
            Concept concept = new Concept();
            concept.setUuid(conceptJson.getAsJsonObject().get(UUID_KEY).getAsString());
            observation.setConcept(concept);
        }
        return observation;
    }

}

File: openmrs-client/src/main/java/org/openmrs/mobile/utilities/ObservationDeserializer.kt
package org.openmrs.mobile.utilities

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import org.openmrs.mobile.models.Concept
import org.openmrs.mobile.models.Observation
import java.lang.reflect.Type


class ObservationDeserializer : JsonDeserializer<Observation> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Observation {
        val jsonObject = json.asJsonObject
        val observation = Observation()
        observation.uuid = jsonObject[UUID_KEY].asString
        observation.display = jsonObject[DISPLAY_KEY].asString
        val conceptJson = jsonObject["concept"]
        if (conceptJson != null && "Visit Diagnoses" == conceptJson.asJsonObject[DISPLAY_KEY].asString) {
            val diagnosisDetailJSONArray = jsonObject["groupMembers"].asJsonArray
            for (i in 0 until diagnosisDetailJSONArray.size()) {
                val diagnosisDetails = diagnosisDetailJSONArray[i].asJsonObject
                val diagnosisDetail = diagnosisDetails["concept"].asJsonObject[DISPLAY_KEY].asString
                if ("Diagnosis order" == diagnosisDetail) {
                    observation.diagnosisOrder = diagnosisDetails.asJsonObject[VALUE_KEY].asJsonObject[DISPLAY_KEY].asString
                } else if ("Diagnosis certainty" == diagnosisDetail) {
                    observation.setDiagnosisCertanity(
                            diagnosisDetails.asJsonObject[VALUE_KEY].asJsonObject[DISPLAY_KEY].asString)
                } else {
                    try {
                        observation.diagnosisList = diagnosisDetails.asJsonObject[VALUE_KEY].asJsonObject[DISPLAY_KEY].asString
                    } catch (e: IllegalStateException) {
                        observation.diagnosisList = diagnosisDetails.asJsonObject[VALUE_KEY].asString
                    }
                }
            }
        } else if (conceptJson != null && "Text of encounter note" == conceptJson.asJsonObject[DISPLAY_KEY].asString) {
            observation.diagnosisNote = jsonObject.asJsonObject[VALUE_KEY].asString
        }
        if (conceptJson != null) {
            val concept = Concept()
            concept.uuid = conceptJson.asJsonObject[UUID_KEY].asString
            observation.concept = concept
        }
        return observation
    }

    companion object {
        private const val UUID_KEY = "uuid"
        private const val DISPLAY_KEY = "display"
        private const val VALUE_KEY = "value"
    }
}
